async-nats/src/service/mod.rs

```diff
@@ -748,3 +748,115 @@ impl Request {
         result
     }
 }
+
+#[derive(Debug)]
+pub struct EndpointBuilder {
+    client: Client,
+    stats: Arc<Mutex<Endpoints>>,
+    shutdown_tx: Sender<()>,
+    name: Option<String>,
+    metadata: Option<HashMap<String, String>>,
+    schema: Option<endpoint::Schema>,
+    subjects: Arc<Mutex<Vec<String>>>,
+}
+
+impl EndpointBuilder {
+    fn new(
+        client: Client,
+        stats: Arc<Mutex<Endpoints>>,
+        shutdown_tx: Sender<()>,
+        subjects: Arc<Mutex<Vec<String>>>,
+    ) -> EndpointBuilder {
+        EndpointBuilder {
+            client,
+            stats,
+            subjects,
+            shutdown_tx,
+            name: None,
+            metadata: None,
+            schema: None,
+        }
+    }
+
+    /// Set the name of the [Endpoint].
+    pub fn name<S: ToString>(mut self, name: S) -> EndpointBuilder {
+        self.name = Some(name.to_string());
+        self
+    }
+
+    /// Adds metadata specific for the [Endpoint].
+    pub fn metadata(mut self, metadata: HashMap<String, String>) -> EndpointBuilder {
+        self.metadata = Some(metadata);
+        self
+    }
+
+    /// Specifies [Schema] for the [Endpoint].
+    pub fn schema<S: ToString>(mut self, request: S, response: S) -> EndpointBuilder {
+        self.schema = Some(endpoint::Schema {
+            request: request.to_string(),
+            response: response.to_string(),
+        });
+        self
+    }
+
+    /// Finalizes the builder and adds the [Endpoint].
+    pub async fn add<S: ToString>(self, subject: S) -> Result<Endpoint, Error> {
+        let subject = subject.to_string();
+        let name = self.name.clone().unwrap_or_else(|| subject.clone());
+        let requests = self
+            .client
+            .queue_subscribe(subject.clone(), QUEUE_GROUP.to_string())
+            .await?;
+        debug!("created service for endpoint {subject}");
+
+        let shutdown_rx = self.shutdown_tx.subscribe();
+
+        let mut stats = self.stats.lock().unwrap();
+        stats
+            .endpoints
+            .entry(subject.clone())
+            .or_insert(endpoint::Inner {
+                name,
+                metadata: self.metadata.unwrap_or_default(),
+                schema: self.schema,
+                ..Default::default()
+            });
+        self.subjects.lock().unwrap().push(subject.clone());
+        Ok(Endpoint {
+            requests,
+            stats: self.stats.clone(),
+            client: self.client.clone(),
+            endpoint: subject.clone(),
+            shutdown: Some(shutdown_rx),
+            shutdown_future: None,
+        })
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct Schema {
+    #[serde(rename = "type")]
+    pub response_type: String,
```

Comment:
Are we calling type kind in jetstream? for paged requests? I believe so.

```suggestion
    pub kind: String,
```